netebpfext/net_ebpf_ext_xdp.c

```diff
@@ -667,42 +667,49 @@ net_ebpf_ext_layer_2_classify(
         net_xdp_ctx.base.data_end = packet_buffer + net_buffer->DataLength;
     }
 
-    if (net_ebpf_extension_hook_invoke_program(attached_client, &net_xdp_ctx, &result) == EBPF_SUCCESS) {
-        switch (result) {
-        case XDP_PASS:
-            if (net_xdp_ctx.cloned_nbl != NULL) {
-                // Drop the original NBL.
-                classify_output->actionType = FWP_ACTION_BLOCK;
-                classify_output->rights &= ~FWPS_RIGHT_ACTION_WRITE;
-
-                // Inject the cloned NBL in receive path.
-                status = _net_ebpf_ext_receive_inject_cloned_nbl(net_xdp_ctx.cloned_nbl, incoming_fixed_values);
-                if (NT_SUCCESS(status))
-                    // If cloned packet could be successfully injected, no need to audit for dropping the original.
-                    // So absorb the original packet.
-                    classify_output->flags |= FWPS_CLASSIFY_OUT_FLAG_ABSORB;
-            }
-            // No special processing required in the non-clone case.
-            // The inbound original NBL will be allowed to proceed in the ingress path.
-            break;
-        case XDP_TX:
-            _net_ebpf_ext_handle_xdp_tx(&net_xdp_ctx, incoming_fixed_values);
-            // Absorb the original NBL.
-            classify_output->actionType = FWP_ACTION_BLOCK;
-            classify_output->rights &= ~FWPS_RIGHT_ACTION_WRITE;
-            classify_output->flags |= FWPS_CLASSIFY_OUT_FLAG_ABSORB;
-            break;
-        case XDP_DROP:
+    if (net_ebpf_extension_hook_invoke_program(attached_client, &net_xdp_ctx, &result) != EBPF_SUCCESS) {
+        // Perform a default action if the program fails.
+        result = XDP_PASS;
```

Comment:
Shouldn't this be XDP_DROP by default, i.e., secure-by-default?

Reply:
Great question, and I tried to sidestep this by simply codifying the current "do-nothing-by-default" behavior. I had a few questions:

1. Is dropping every packet, which effectively takes the machine off the network, more secure than allowing all packets through to the machine? Is this a self-DoS?
2. What if the XDP program isn't intended for DDoS protection, e.g. a simple packet monitor, and the system may start experiencing transient packet loss with eBPF + XDP? Is this going to be disruptive, especially if the program may "fail" for nontrivial time intervals?
3. Should eBPF itself take responsibility for providing a default result if the program fails? In other words, should eBPF mask any internal failure and obviate the need for this check within a hook/extension? This doesn't affect _what_ action is returned, but moving this check to eBPF allows the default policy to be decoupled from the hook implementation.

Reply:
Looking at the code, even though `net_ebpf_extension_hook_invoke_program` returns an error code (look at `_ebpf_link_instance_invoke`), it looks like should not fail in normal conditions. Also, we acquire rundown reference before invoking the program, so it is guaranteed that the program is not getting detached. In the case `net_ebpf_extension_hook_invoke_program` fails, I think we should drop the packet since there is an eBPF program attached to the hook, which execution context is failing to invoke.

Reply:
> Is dropping every packet, which effectively takes the machine off the network, more secure than allowing all packets through to the machine?

Definitely more secure if there is no network access.

> Should eBPF itself take responsibility for providing a default result if the program fails?

I agree with Anurag's answer.

Reply:
Fair enough; I've updated the default action to `XDP_DROP` and will take the same action within xdp-for-windows.